media-proxy/src/rdma_hmem.c

```diff
@@ -0,0 +1,158 @@
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ */
+
+#if HAVE_CONFIG_H
+    #include <config.h>
+#endif
+
+#include <inttypes.h>
+#include <stdbool.h>
+#include "rdma_hmem.h"
+
+static bool hmem_initialized = false;
+
+struct rdma_hmem_ops {
+    int (*init)(void);
+    int (*cleanup)(void);
+    int (*alloc)(uint64_t device, void **buf, size_t size);
+    int (*alloc_host)(void **buf, size_t size);
+    int (*free)(void *buf);
+    int (*free_host)(void *buf);
+    int (*mem_set)(uint64_t device, void *buf, int value, size_t size);
+    int (*copy_to_hmem)(uint64_t device, void *dst, const void *src,
+                size_t size);
+    int (*copy_from_hmem)(uint64_t device, void *dst, const void *src,
+                  size_t size);
+    int (*get_dmabuf_fd)(void *buf, size_t len,
+                 int *fd, uint64_t *offset);
+};
+
+static struct rdma_hmem_ops hmem_ops[] = {
+    [FI_HMEM_SYSTEM] = {
+        .init = rdma_host_init,
+        .cleanup = rdma_host_cleanup,
+        .alloc = rdma_host_alloc,
+        .alloc_host = rdma_default_alloc_host,
+        .free = rdma_host_free,
+        .free_host = rdma_default_free_host,
+        .mem_set = rdma_host_memset,
+        .copy_to_hmem = rdma_host_memcpy,
+        .copy_from_hmem = rdma_host_memcpy,
+        .get_dmabuf_fd = rdma_hmem_no_get_dmabuf_fd,
+    },
+    // More devices could be added
+};
+
+int rdma_hmem_init(enum fi_hmem_iface iface)
+{
+    int ret;
+
+    ret = hmem_ops[iface].init();
+    if (ret == FI_SUCCESS)
+        hmem_initialized = true;
+
+    return ret;
+}
+
+int rdma_hmem_cleanup(enum fi_hmem_iface iface)
+{
+    int ret = FI_SUCCESS;
+
+    if (hmem_initialized) {
+        ret = hmem_ops[iface].cleanup();
+        if (ret == FI_SUCCESS)
+            hmem_initialized = false;
+    }
+
+    return ret;
+}
+
+int rdma_hmem_alloc(enum fi_hmem_iface iface, uint64_t device, void **buf,
+          size_t size)
+{
+    return hmem_ops[iface].alloc(device, buf, size);
+}
+
+int rdma_default_alloc_host(void **buf, size_t size)
+{
+    *buf = malloc(size);
+    return (*buf == NULL) ? -FI_ENOMEM : 0;
+}
+
+int rdma_default_free_host(void *buf)
+{
+    free(buf);
+    return 0;
+}
```

Comment:
Calling `free()` against a NULL pointer is okay by its design. Reference: https://linux.die.net/man/3/free
> The free() function frees the memory space pointed to by ptr, which must have been returned by a previous call to malloc(), calloc() or realloc(). Otherwise, or if free(ptr) has already been called before, undefined behavior occurs. If ptr is NULL, no operation is performed.

Reply:
Thanks for clarifying. But then should we call `buf=NULL` befor return?

Reply:
It could be argued that NULLing the pointer is preferable, but as the function takes a single pointer, it cannot do this, so the user will not expect it. I left it as it is, because it is not even used now.

When and if we will use rdma_hmem.c for example to allocate memory on GPU we can redesign it little bit better

Reply:
Agree